turbopack/crates/turbo-persistence/src/compaction/interval_map.rs

```diff
@@ -0,0 +1,166 @@
+struct IntervalPoint<T> {
+    start: u64,
+    value: Option<T>,
+}
+
+pub struct IntervalMap<T> {
+    intervals: Vec<IntervalPoint<T>>,
+}
+
+impl<T> Default for IntervalMap<T> {
+    fn default() -> Self {
+        Self {
+            intervals: Vec::new(),
+        }
+    }
+}
+
+impl<T> IntervalMap<T> {
+    pub fn new() -> Self {
+        Default::default()
+    }
+
+    /// Ensures that there is an interval point at the specified location.
+    fn ensure_point(&mut self, location: u64) -> usize
+    where
+        T: Clone,
+    {
+        match self
+            .intervals
+            .binary_search_by_key(&location, |point| point.start)
+        {
+            Ok(index) => index,
+            Err(index) => {
+                // If the location does not exist, we need to insert a new interval
+                let value = if index > 0 {
+                    self.intervals[index - 1].value.clone()
+                } else {
+                    None
+                };
+                self.intervals.insert(
+                    index,
+                    IntervalPoint {
+                        start: location,
+                        value,
+                    },
+                );
+                index
+            }
+        }
+    }
+
+    /// Applies the update function to all values in the specified range.
+    pub fn update(&mut self, range: &(u64, u64), mut update: impl FnMut(&mut T))
+    where
+        T: Default + Clone,
+    {
+        let start = range.0;
+        let end = range.1;
+        if start > end {
+            return;
+        }
+
+        let start = self.ensure_point(start);
```

Comment:
mutating the range doesn't seem required here,  just binary search for the endpoints like in `test`?

Reply:
i suppose after seeing the use case i think i get it.  perhaps `upsert` is a better name?

Reply:
Conceptionally this struct is like a more efficient `[T; u64::MAX]`. This method updates items from start to end.

The internal representation is a list of intervals. It's an implementation detail that this operation need to split the intervals.